pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go

```diff
@@ -241,13 +224,66 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	return ctrl.Result{}, r.UpdateStatus(ctx, app)
 }
 
-// if any finalizers newly registered, return true
-func registerFinalizers(app *v1beta1.Application) bool {
-	if !meta.FinalizerExists(&app.ObjectMeta, resourceTrackerFinalizer) && app.Status.ResourceTracker != nil {
-		meta.AddFinalizer(&app.ObjectMeta, resourceTrackerFinalizer)
-		return true
+func (r *Reconciler) handleFinalizers(ctx context.Context, app *v1beta1.Application) (bool, error) {
+	if app.ObjectMeta.DeletionTimestamp.IsZero() {
+		// NOTE Because resource tracker is cluster-scoped resources, we cannot garbage collect them
+		// by setting application(namespace-scoped) as their owner. So we must delete all
+		// resource trackers through app controller's finalizer logic.
+
+		// 'resourceTrackerFinalizer' is used to delete the resource tracker of the last app revision
+		if !meta.FinalizerExists(&app.ObjectMeta, resourceTrackerFinalizer) {
+			meta.AddFinalizer(&app.ObjectMeta, resourceTrackerFinalizer)
+			klog.InfoS("Register new finalizer for application", "application", klog.KObj(app), "finalizer", resourceTrackerFinalizer)
+			return true, errors.Wrap(r.Client.Update(ctx, app), errUpdateApplicationFinalizer)
+		}
+		// 'onlyRevisionFinalizer' is used to delete all resource trackers of app revisions which
+		// may be used out of the domain of app controller, e.g., AppRollout controller.
+		if app.Annotations[oam.AnnotationAppRevisionOnly] == "true" ||
+			len(app.Annotations[oam.AnnotationAppRollout]) != 0 || app.Spec.RolloutPlan != nil {
+			if !meta.FinalizerExists(&app.ObjectMeta, onlyRevisionFinalizer) {
+				meta.AddFinalizer(&app.ObjectMeta, onlyRevisionFinalizer)
+				klog.InfoS("Register new finalizer for application", "application", klog.KObj(app), "finalizer", onlyRevisionFinalizer)
+				return true, errors.Wrap(r.Client.Update(ctx, app), errUpdateApplicationFinalizer)
+			}
+		}
+	} else {
+		if meta.FinalizerExists(&app.ObjectMeta, resourceTrackerFinalizer) {
+			if app.Status.LatestRevision != nil && len(app.Status.LatestRevision.Name) != 0 {
+				latestTracker := &v1beta1.ResourceTracker{}
+				latestTracker.SetName(dispatch.ConstructResourceTrackerName(app.Status.LatestRevision.Name, app.Namespace))
+				if err := r.Client.Delete(ctx, latestTracker); err != nil && !kerrors.IsNotFound(err) {
+					klog.ErrorS(err, "Failed to delete latest resource tracker", "name", latestTracker.Name)
+					app.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, "error to  remove finalizer")))
+					return true, errors.Wrap(r.UpdateStatus(ctx, app), errUpdateApplicationStatus)
+				}
+			}
+			meta.RemoveFinalizer(app, resourceTrackerFinalizer)
+			return true, errors.Wrap(r.Client.Update(ctx, app), errUpdateApplicationFinalizer)
+		}
+		if meta.FinalizerExists(&app.ObjectMeta, onlyRevisionFinalizer) {
```

Comment:
do these two finalizer conflicts? Why do we need two different logic?

Reply:
they are not conflict. but `onlyRevisionFinalizer` has to do `list` operation while `resourceTrackerFinaler`just need `get`.   It tries to avoid doing `list` operation every time.